ibis/backends/pandas/client.py

```diff
@@ -192,37 +195,131 @@ def convert_any_to_timestamp(_, out_dtype, column):
             return column.dt.tz_localize(timezone)
 
 
-@sch.convert.register(object, dt.Struct, pd.Series)
-def convert_struct_to_dict(_, out_dtype, column):
-    def convert_element(values, names=out_dtype.names):
-        if values is None or isinstance(values, dict) or pd.isna(values):
-            return values
-        return dict(zip(names, values))
+@sch.convert.register(object, dt.DataType, (type(None), type(pd.NA)))
+@sch.convert.register(object, dt.Boolean, bool)
+@sch.convert.register(object, dt.String, str)
+@sch.convert.register(object, dt.Integer, (int, np.integer))
+@sch.convert.register(object, dt.Floating, float)
+@sch.convert.register(object, dt.Date, datetime.date)
+@sch.convert.register(object, dt.Timestamp, datetime.datetime)
+@sch.convert.register(object, dt.Time, datetime.time)
+@sch.convert.register(object, dt.Interval, datetime.timedelta)
+@sch.convert.register(object, dt.JSON, (list, dict, int, float))
+def scalar_no_op(_, out_dtype, v):
+    return v
+
+
+@sch.convert.register(object, dt.Array, (list, np.ndarray, tuple))
+def scalar_array(in_, out_dtype, v):
+    try:
+        return list(map(partial(sch.convert, in_, out_dtype.value_type), v))
+    except NotImplementedError:
+        # we have to handle postgres arrays, which always appear as array<T>
+        # even if `v` is a list of lists. This is a fundamental limitation of
+        # postgres arrays: the dimension information is lost upon table
+        # creation
+        #
+        # this recurses over the input elements, bottoming out `try:` body
+        # succeeds i.e., when all elements are no longer lists
+        return list(map(partial(sch.convert, in_, out_dtype), v))
+
+
+@sch.convert.register(object, dt.Map, dict)
+def scalar_map(in_, out_dtype, v):
+    convert_key = partial(sch.convert, in_, out_dtype.key_type)
+    convert_value = partial(sch.convert, in_, out_dtype.value_type)
+    return {convert_key(key): convert_value(value) for key, value in v.items()}
+
+
+@sch.convert.register(object, dt.Struct, dict)
+def scalar_struct(in_, out_dtype, v):
+    return {key: sch.convert(in_, out_dtype[key], value) for key, value in v.items()}
+
+
+@sch.convert.register(object, dt.Floating, (int, decimal.Decimal, str, np.integer))
+def scalar_float_convert(_, out_dtype, v):
+    return float(v)
+
+
+@sch.convert.register(object, dt.Timestamp, str)
+def scalar_timestamp_convert(_, out_dtype, v):
+    return date_parse(v)
+
 
-    return column.map(convert_element)
+@sch.convert.register(object, dt.Integer, float)
+def scalar_int_convert(_, out_dtype, v):
+    if math.isnan(v):
+        return None
+    return int(v)
+
+
+@sch.convert.register(np.dtype, (dt.JSON, dt.Array, dt.Map, dt.Struct), str)
+def convert_json_to_scalar(in_, out, value: str):
```

Comment:
Do we really want to enable this conversion for all backends? IIUC this is to needed since snowflake returns all nested types as JSON strings - could we handle this in a separate external mechanism instead of in `Schema.apply_to`?

Reply:
This is done per backend now!